vllm/spec_decode/spec_decode_worker.py

```diff
@@ -612,22 +631,31 @@ def _create_output_sampler_list(
                             step_index][sequence_index],
                         token_id_logprob=accepted_token_id_logprobs_by_step[
                             step_index][sequence_index],
-                        seq_id=seq_ids[sequence_index],
+                        seq_id=seq_id,
                         topk_token_ids=topk_indices_by_step[step_index]
                         [sequence_index][:num_logprobs],
                         topk_logprobs=topk_logprobs_by_step[step_index]
                         [sequence_index][:num_logprobs],
                     ))
-
             sampler_output_list.append(
                 SamplerOutput(outputs=step_output_token_ids))
+        if self.proposer_uses_kv_cache:
+            # The proposer worker uses KV cache. Hence populate the data
+            # structures needed to keep track of bonus tokens.
+            for seq_index, seq_id in enumerate(seq_ids):
+                last_token_id = accepted_token_ids_by_step[-1][seq_index]
+                if last_token_id == -1:
+                    self.seq_with_bonus_token_in_last_step.discard(seq_id)
+                else:
+                    self.seq_with_bonus_token_in_last_step.add(seq_id)
+            for request_id, seq_ids in request_ids_seq_ids_mapping.items():
+                self.request_id_seq_id_mapping[request_id].update(seq_ids)
```

Comment:
nit: suggest moving to own function so someone reading this function doesn't have to know about all these details